rolo/serving/twisted.py

```diff
@@ -0,0 +1,422 @@
+
+
+class HeaderPreservingHTTPChannel(HTTPChannel):
+    """
+    Special HTTPChannel implementation that uses ``Headers._caseMappings`` to retain header casing both for
+    request headers (server -> WSGI), and  response headers (WSGI -> client).
+    """
+
+    requestFactory = TwistedRequestAdapter
+
+    @staticmethod
+    def protocol_factory():
+        return _GenericHTTPChannelProtocol(HeaderPreservingHTTPChannel())
+
+    def headerReceived(self, line):
+        if not super().headerReceived(line):
+            return False
+
+        # remember casing of headers for requests
+        header, data = line.split(b":", 1)
+        request: TwistedRequestAdapter = self.requests[-1]
+        request.requestHeaders._caseMappings[header.lower()] = header
+        return True
+
+    def writeHeaders(self, version, code, reason, headers):
+        """Alternative implementation that writes the raw headers instead of sanitized versions."""
+        responseLine = version + b" " + code + b" " + reason + b"\r\n"
+        headerSequence = [responseLine]
+
+        for name, value in headers:
+            line = name + b": " + value + b"\r\n"
+            headerSequence.append(line)
+
+        headerSequence.append(b"\r\n")
+        self.transport.writeSequence(headerSequence)
+
+    def isSecure(self):
+        # used to determine the WSGI url scheme (http vs https)
+        try:
+            # ``self.transport`` will be a ``TLSMultiplexer`` instance in our case
+            return self.transport.isSecure()
+        except AttributeError:
+            return super().isSecure()
+
+
+class HeaderPreservingWSGIResponse(_WSGIResponse):
+    def __init__(self, reactor, threadpool, application, request):
+        super().__init__(reactor, threadpool, application, request)
+        update_wsgi_environment(self.environ, request)
+
+    def startResponse(self, *args, **kwargs):
+        result = super().startResponse(*args, **kwargs)
+        # before starting the WSGI response, make sure we store the raw case mappings into the response
+        # headers
+        for header, _ in self.headers:
+            header = header.encode("latin-1")
+            self.request.responseHeaders._caseMappings[header.lower()] = header
+        return result
+
+
+class HeaderPreservingWSGIResource(WSGIResource):
+    def render(self, request):
+        response = HeaderPreservingWSGIResponse(
+            self._reactor, self._threadpool, self._application, request
+        )
+        response.start()
+        return NOT_DONE_YET
+
+
+@implementer(IResource)
+class WebsocketResourceDecorator(proxyForInterface(IResource)):
+    original: WSGIResource
+    isLeaf = True
+
+    def __init__(
+        self,
+        original: WSGIResource,
+        websocketListener: WebSocketListener,
+    ):
+        super().__init__(original)
+        self.websocketListener = websocketListener
+        self.channel = None
```

Comment:
q: should we keep a reference to a channel here? Wouldn't this be overwritten if we get 2 concurrent websockets connections? or is this class instantiated for each incoming request?

Reply:
really nice catch! obviously this was nonsensical. cleaned it up to not hold the channel at all.